### PR TITLE
Refactor rdkit.ML.Cluster.Butina to use numpy to perform on large datasets.

### LINKED ISSUE #6166: Butima clustering fails due to lack of RAM on large datasets.
When I run `rdkit.ML.Cluster.Butima` on a dataset of over 100k molecules it runs out of memory and crashes.

I'm making this issues b/c I already have a PR that allows it to run, and want to reference an issue.  I know that it is an issue for others as well b/c the first web pages a person finds on Google that refer to molecule clustering show attempts to use this that fail:
https://www.macinchem.org/reviews/clustering/clustering.php

PR to follow with narrowly focused changes to reduce memory usage...

### PR DESCRIPTION
**Pull request by sakoht**
#### Reference Issue
Fixes #6166

#### What does this implement/fix? Explain your changes.
I attempted to use the existing Butina logic to cluster 99k molecules, and it eventually runs out of memory on a 16GB of RAM machine.

With these changes it runs successfully on a dataset of that size.

Luckily this is <60 lines of pure Python that don't use anything else in the system.  It passes your unit test, and also has produced identical results to the old code on our internal datasets.

Hope this is useful to others.

This replaces the list-of-lists with an `numpy.array(dtype=bool)` to track which pairs of molecules pass the threshold, which is smaller.  It creates the list empty, but full-size at the beginning and the loops of code flip the bits instead of grow the list.

### CODE REVIEW

File: rdkit/ML/Cluster/Butina.py [hunk 1]
```diff
@@ -55,14 +55,15 @@ def ClusterData(data, nPts, distThresh, isDistData=False, distFunc=EuclideanDist
...
   if isDistData and len(data) > (nPts * (nPts - 1) / 2):
     logger.warning("Distance matrix is too long")
-  nbrLists = [None] * nPts
-  for i in range(nPts):
-    nbrLists[i] = []
 
+  # Create an empty matrix of booleans with a point for each i,j pair of molecule numbers.
+  # Also tally the count of all pairs, skipping duplicates, and all self-comparison.
+  # For efficiency, we never change the size of this bit bit-table.  We just flip bits.
+  matrix_passing_threshold: numpy.array = numpy.zeros([nPts, nPts], dtype=bool)
```
Comment (greglandrum) [inline rdkit/ML/Cluster/Butina.py:65]: The RDKit doesn't use type hints. Please remove them

**Review by greglandrum — CHANGES_REQUESTED** [review]
Thanks for the contribution!

I can see that this is going to be a win (and a big one) in a lot of circumstances. However, this uses $N^2$ memory. That's going to be a problem for very large datasets. The current approach uses a lot of memory for the case where compounds have a lot of neighbors, but will actually handle very large datasets which are sparse (where compounds have a small number of neighbors) pretty efficiently as far as memory usage is concerned.

Given that, I'd suggest including both implementations and adding another additional parameter to the function which controls which one gets invoked. We can call this something like `useSparseApproach` and have it default to `False` (so your new approach would be selected by default).

Some general style comments to make this match the rest of the RDKit code:
- we use camelCase, not snake_case, please update the code accordingly
- we don't use type hints, please remove them

File: rdkit/ML/Cluster/Butina.py (hunk 1 above)
Comment (sakoht) [reply rdkit/ML/Cluster/Butina.py:65]: I'll update to camelCase too.

**Comment by sakoht** [comment]
Notes on performance: 

This one runs on my MacBook Air M1 laptop w/ 98k small molecules in about 6 hours.  The memory footprint was, I believe 6GB?  I to hold the input cluster data (5 billion datapoints), plus I believe about 1G for the NxN membership matrix, but I'll confirm and post it.  

My guess is that 2x the data will mean 4x the RAM consumed, so somewhere between 100k and 200k it stops working on a modern laptop, and would need to go sparse if a big RAM box isn't available.

**Comment by ilkersigirci** [comment]
Thank you for this. What is the current status of it?

**Comment by greglandrum** [comment]
@sakoht : any plans to take this up again?

**Comment by j-adamczyk** [comment]
@greglandrum how about refactoring this to use SciPy sparse array? It would solve both problems, since sparse arrays allocate memory only for non-zero elements. It would introduce the SciPy as a dependency, although many projects utilizing RDKit in Python already use it as e.g. scikit-learn dependency.

**Comment by sakoht** [comment]
> @sakoht : any plans to take this up again?

@greglandrum, @ilkersigirci Sorry I didn't see your messages way back last Summer.  I meant to circle back and make the requested changes to support both paths of logic and repo standards, but have been swamped with work, so we just used my branch.

Glad to go through and update things now that I have a bit more time.

**Comment by sakoht** [comment]
> @greglandrum how about refactoring this to use SciPy sparse array? It would work in both cases, since sparse arrays allocate memory only for non-zero elements. It would introduce the SciPy as a dependency, although many projects utilizing RDKit in Python already use it as e.g. scikit-learn dependency.


@greglandrum that sounds like a clever idea.  Since we would be letting the caller specify sparse/dense flag and going through different code paths, this makes the sparse code path simpler, and a shared code path will be less likely to diverge.

I think the key question would be what the project maintainers would prefer?

**Comment by greglandrum** [comment]
@sakoht, I would prefer to get this one merged and then discuss adding an optional path through the code for sparse arrays. I don't want to add a hard scipy dependency, but we could definitely have it be optional.

**Pull request closed without merging**